state/protocol/protocol_state/state/mutable_protocol_state_test.go

```diff
@@ -0,0 +1,650 @@
+		s.kvStateMachineFactories[i] = *protocol_statemock.NewKeyValueStoreStateMachineFactory(s.T())
+		s.kvStateMachineFactories[i].On("Create", s.candidate.View, s.candidate.ParentID, &s.parentState, &s.evolvingState).Return(&s.kvStateMachines[i], nil)
+		kvStateMachineFactories[i] = &s.kvStateMachineFactories[i]
+	}
+
+	s.mutableState = newMutableProtocolState(
+		&s.epochProtocolStateDB,
+		&s.protocolKVStoreDB,
+		&s.globalParams,
+		&s.headersDB,
+		&s.resultsDB,
+		kvStateMachineFactories,
+	)
+}
+
+// testEvolveState is the main logic for testing `EvolveState`. Specifically, we test:
+//   - we _always_ require a deferred db update that indexes the protocol state by the candidate block's ID
+//   - we expect a deferred db update that persists the protocol state if and only if there was a state change compared to the parent protocol state
+//
+// Note that the `MutableProtocolState` bundles all deferred database updates into a `DeferredBlockPersist`. Conceptually, it is possible that
+// the `MutableProtocolState` wraps the deferred database operations in faulty code, such that they are eventually not executed. Therefore,
+// we explicitly test here whether the storage functors *generated* by `ProtocolKVStore.IndexTx` and “ProtocolKVStore.StoreTx` are
+// actually called when executing the returned `DeferredBlockPersist`
+func (s *StateMutatorSuite) testEvolveState(seals []*flow.Seal, expectedResultingStateID flow.Identifier, stateChangeExpected bool) {
+	// on the happy path, we _always_ require a deferred db update, which indexes the protocol state by the candidate block's ID
+	indexTxDeferredUpdate := storagemock.NewDeferredDBUpdate(s.T())
+	indexTxDeferredUpdate.On("Execute", mock.Anything).Return(nil).Once()
+	s.protocolKVStoreDB.On("IndexTx", s.candidate.ID(), expectedResultingStateID).Return(indexTxDeferredUpdate.Execute).Once()
+
+	// expect calls to prepare a deferred update for indexing and storing the resulting state:
+	// as state has not changed, we expect the parent blocks protocol state ID
+	storeTxDeferredUpdate := storagemock.NewDeferredDBUpdate(s.T())
+	if stateChangeExpected {
+		storeTxDeferredUpdate.On("Execute", mock.Anything).Return(nil).Once()
+		s.protocolKVStoreDB.On("StoreTx", expectedResultingStateID, &s.evolvingState).Return(storeTxDeferredUpdate.Execute).Once()
+	}
+
+	resultingStateID, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, seals)
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), expectedResultingStateID, resultingStateID)
+
+	// Provide the blockID and execute the resulting `DeferredDBUpdate`. Thereby,
+	// the expected mock methods should be called, which is asserted by the testify framework
+	err = dbUpdates.Pending().WithBlock(s.candidate.ID())(&transaction.Tx{})
+	require.NoError(s.T(), err)
+
+	s.protocolKVStoreDB.AssertExpectations(s.T())
+	indexTxDeferredUpdate.AssertExpectations(s.T())
+	storeTxDeferredUpdate.AssertExpectations(s.T())
```

Comment:
```suggestion
```

Pretty sure it can be omitted since `AssertExpectations` is called automatically when test finishes.

Reply:
My understanding is that `AssertExpectations` is called when the test finishes, but _not_ when the sub-test finishes. However, in our case, we have specific expectations for each of our sub-tests. 

Consider the following example (you can just copy-paste it into  `mutable_protocol_state_test.go`).  As the example shows, the test passes if the expected call is made in _any_ of the sub-tests. Since we are calling `EvolveState` in each sub-test, I find it very hard to reason that the expectations are satisfied by the specific sub-tests and not accidentally somewhere else. 

```golang
func (s *StateMutatorSuite) Test_AssertExpectations() {
	storeTxDeferredUpdate := storagemock.NewDeferredDBUpdate(s.T())

	s.Run("expect `storeTxDeferredUpdate.Execute` to be called", func() {
		storeTxDeferredUpdate.On("Execute", mock.Anything).Return(nil).Once()

		// tested business logic does _not_ make the expected call

		// unless we call AssertExpectations in the sub-test, the sub-test will not detect the missing call
		//storeTxDeferredUpdate.AssertExpectations(s.T())
	})

	s.Run("expect `storeTxDeferredUpdate.Execute` *not* to be called", func() {
		// tested business logic cales Execute even though it is not supposed to
		storeTxDeferredUpdate.Execute(&transaction.Tx{})
	})
}
```

I totally admit that the example in its shortness is a bit academic and I also can't point to any specific scenario in our tests, where something would go wrong without the `AssertExpectations` calls. So if you have a conclusive argument why I don't have to worry about missing any expectation in the sub-test, totally happy to remove the  `AssertExpectations` calls. But in absence of such an argument, I would be inclined to leave these `AssertExpectations` - after all its not a significant addition of complexity.

Reply:
for now, I have just included the following documentation: https://github.com/onflow/flow-go/blob/13b938503aeaa079e4798ce392d4796899d3262f/state/protocol/protocol_state/state/mutable_protocol_state_test.go#L126-L133